docs/rfcs/2023-01-03-procedure-framework.md

```diff
@@ -0,0 +1,151 @@
+---
+Feature Name: "procedure-framework"
+Tracking Issue: https://github.com/GreptimeTeam/greptimedb/issues/286
+Date: 2023-01-03
+Author: "Yingwen <realevenyag@gmail.com>"
+---
+
+Procedure Framework
+----------------------
+
+# Summary
+A framework for executing operations in a fault-tolerant manner.
+
+# Motivation
+Some operations in GreptimeDB require multiple steps to implement. For example, creating a table needs:
+1. Check whether the table exists
+2. Create the table in the table engine
+  1. Create a region for the table in the storage engine
+  2. Persist the metadata of the table to the table manifest
+3. Add the table to the catalog manager
+
+If the node dies or restarts in the middle of creating a table, it could leave the system in an inconsistent state. The procedure framework, inspired by [Apache HBase's ProcedureV2 framework](https://github.com/apache/hbase/blob/bfc9fc9605de638785435e404430a9408b99a8d0/src/main/asciidoc/_chapters/pv2.adoc) and [Apache Accumulo’s FATE framework](https://accumulo.apache.org/docs/2.x/administration/fate), aims to provide a unified way to implement multi-step operations that is tolerant to failure.
+
+# Details
+## Overview
+The procedure framework consists of the following primary components:
+- A `Procedure` represents an operation or a set of operations to be performed step-by-step
+- `ProcedureManager`, the system to run `Procedures`
+- `ProcedureStore` is a storage layer for persisting the procedure state
+
+
+## Procedures
+The `ProcedureManager` keeps calling `Procedure::execute()` until the Procedure is done, so the operation of the Procedure should be [idempotent](https://developer.mozilla.org/en-US/docs/Glossary/Idempotent): it needs to be able to undo or replay a partial execution of itself.
+
+```rust
+trait Procedure {
+    fn execute(&mut self, ctx: &Context) -> Result<Status>;
+
+    fn dump(&self) -> Result<String>;
+
+    // other methods...
+}
+```
+
+The `Status` is an enum that has the following variants:
+```rust
+enum Status {
+    Executing {
+        persist: bool,
+    },
+    Suspended {
+        subprocedures: Vec<ProcedureWithId>,
+        persist: bool,
+    },
+    Done,
+}
+```
+
+A call to `execute()` can result in the following possibilities:
+- `Ok(Status::Done)`: we are done
+- `Ok(Status::Executing { .. })`: there are remaining steps to do
+- `Ok(Status::Suspend { .. })`: execution is suspended and can be resumed later
+- `Err(e)`: error occurs during execution and the procedure is unable to proceed anymore.
+
+The framework assigns a unique id for each procedure and the procedure can get this id via the `Context`.
+
+```rust
+struct Context {
+    id: ProcedureId,
```

Comment:
Would it be better to make ID an attribute field of one procedure? I think this is a unique, context-free and immutable field.

Reply:
I prepare to add a `procedure_id` method to the Procedure trait.

Reply:
I found this makes the parent procedure difficult to assign a procedure id to its subprocedure returned by other methods. The parent procedure needs to assign the same procedure id to its sub-procedures so we can load the sub-procedure's state from the procedure store.

https://github.com/evenyag/greptimedb/blob/66816c1f1048725002c978abbaea4a44c8ec6b0a/src/datanode/src/procedure/create.rs#L269-L283

What about always requiring callers to generate the procedure id manually? We can change the `ProcedureManager::submit` to take a `ProcedureWithId`.

Reply:
>The parent procedure needs to assign the same procedure id to its sub-procedures

Can I read this as "parent procedure share its ID with sub-procedures"? Then what does this means https://github.com/GreptimeTeam/greptimedb/pull/836#discussion_r1089711943

>What about always requiring callers to generate the procedure id manually? We can change the ProcedureManager::submit to take a ProcedureWithId.

Or we can put an `IdAllocator` into context, which allocates new id for root procedure, and returns parent's id for sub-procedures.

Reply:
> The parent procedure needs to assign the same procedure id to its sub-procedures

This means the parent procedure generates the same procedure id for the sub-procedure after the process is restarted. So the procedure manager can use the id to load the sub-procedure from the procedure store and skips the steps already executed.